File: backend/src/main/java/com/rungo/api/domain/marathon/course/entity/Course.java
package com.rungo.api.domain.marathon.course.entity;

import com.rungo.api.domain.marathon.course.status.CourseStatus;
import com.rungo.api.domain.marathon.marathon.entity.Marathon;
import com.rungo.api.global.exception.CustomException;
import com.rungo.api.global.exception.ErrorCode;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;

@Entity
@Getter
@NoArgsConstructor
public class Course {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Setter
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "marathon_id", nullable = false)
    private Marathon marathon;

    @Column(nullable = false)
    private String courseType;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    @Column(nullable = false)
    private Integer capacity;

    @Column(nullable = false)
    private Integer currentCount;

    public Course(
            String courseType,
            BigDecimal price,
            Integer capacity,
            Integer currentCount
    ) {
        this.courseType = courseType;
        this.price = price;
        this.capacity = capacity;
        this.currentCount = currentCount;
    }

    public CourseStatus getStatus() {
        return currentCount >= capacity ? CourseStatus.FULL : CourseStatus.AVAILABLE;
    }
    public boolean isFull() {
        return currentCount >= capacity;
    }

    public void resetCurrentCount() {
        this.currentCount = 0;
    }

    public Integer getRemainingCount(){
        return this.capacity - this.currentCount;
    }

    public void updateCourseInfo(
            String courseType,
            BigDecimal price,
            Integer capacity
    ) {
        if (courseType != null) this.courseType = courseType;
        if (price != null) this.price = price;
        if (capacity != null) this.capacity = capacity;
    }
}

File: backend/src/main/java/com/rungo/api/domain/marathon/course/entity/Course.kt
package com.rungo.api.domain.marathon.course.entity

import com.rungo.api.domain.marathon.course.status.CourseStatus
import com.rungo.api.domain.marathon.marathon.entity.Marathon
import jakarta.persistence.*
import java.math.BigDecimal

@Entity
class Course protected constructor() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0L
        protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "marathon_id", nullable = false)
    lateinit var marathon: Marathon
        protected set

    @Column(nullable = false)
    lateinit var courseType: String
        protected set

    @Column(nullable = false, precision = 10, scale = 2)
    lateinit var price: BigDecimal
        protected set

    @Column(nullable = false)
    var capacity: Int = 0
        protected set

    @Column(nullable = false)
    var currentCount: Int = 0
        protected set

    fun assignMarathon(marathon: Marathon) {
        this.marathon = marathon
    }

    fun getStatus(): CourseStatus {
        return if (currentCount >= capacity) {
            CourseStatus.FULL
        } else {
            CourseStatus.AVAILABLE
        }
    }

    fun isFull(): Boolean {
        return currentCount >= capacity
    }

    fun resetCurrentCount() {
        currentCount = 0
    }

    fun getRemainingCount(): Int {
        return capacity - currentCount
    }

    fun updateCourseInfo(
        courseType: String?,
        price: BigDecimal?,
        capacity: Int?
    ) {
        if (courseType != null) this.courseType = courseType
        if (price != null) this.price = price
        if (capacity != null) this.capacity = capacity
    }

    companion object {

        @JvmStatic
        fun create(
            courseType: String,
            price: BigDecimal,
            capacity: Int,
            currentCount: Int
        ): Course {

            return Course().apply {
                this.courseType = courseType
                this.price = price
                this.capacity = capacity
                this.currentCount = currentCount
            }
        }
    }
}